002-es6-modules.md

```diff
@@ -0,0 +1,441 @@
+//   etc.
+import '/bar';
+```
+
+```javascript
+// continues to *search*:
+//   ./node_modules/baz.js
+//   ./node_modules/baz/package.json
+//   ./node_modules/baz/index.js
+// and parent node_modules:
+//   ../node_modules/baz.js
+//   ../node_modules/baz/package.json
+//   ../node_modules/baz/index.js
+//   etc.
+import 'baz';
+```
+
+#### Removal of non-local dependencies
+
+All of the following:
+
+* `$NODE_PATH`
+* `$HOME/.node_modules`
+* `$HOME/.node_libraries`
+* `$PREFIX/lib/node`
+
+will not be supported by the `import` statement. Use local dependencies, and symbolic links as needed.
+
+##### How to support non-local dependencies
+
+Although not recommended, and in fact discouraged, there is a way to support non-local dependencies. **USE THIS AT YOUR OWN DISCRETION**. 
+
+Symlinks of `node_modules -> $HOME/.node_modules`, `node_modules/foo/ -> $HOME/.node_modules/foo/`, etc. will continue to be supported.
+
+Adding a parent directory with `node_modules` symlinked will be an effective strategy for recreating these functionalities. This will incur the known problems with non-local dependencies, but now leaves the problems in the hands of the user, allowing node to give more clear insight to your modules by reducing complexity.
+
+Given:
+
+```sh
+/opt/local/myapp
+```
+
+Transform to:
+
+```sh
+/opt/local/non-local-deps/myapp
+/opt/local/non-local-deps/node_modules -> $PREFIX/lib/node (etc.)
+```
+
+And nest as many times as needed.
+
+#### Errors
+
+In the case that an `import` statement is unable to find a module, node should make a **best effort** to see if `require` would have found the module and print out where it was found, if `NODE_PATH` was used, if `HOME` was used, etc.
+
+#### Vendored modules
+
+This will mean vendored modules are not included in the search path since `package.json` is not searched for outside of `node_modules`. Please use [bundledDependencies](https://docs.npmjs.com/files/package.json#bundleddependencies) to vendor your dependencies instead.
+
+#### Shipping both ES6 and CJS
+
+Since `node_modules` continues to use searching, when a `package.json` main is encountered we are still able to perform file extension searches. If we have 2 entry points `index.jsm` and `index.js` by setting `main:"./index"` we can let Node pick up either depending on what is supported, without us needing to manage multiple entry points separately.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`. `require` will not automatically wrap ES6 modules in a `Promise`. In the future if top level `await` becomes spec, you can use the `System.loader.import` function to wrap modules and wait on them (top level await can cause deadlock with circular dependencies, node should discourage its use).
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+const foo = require('./es6').foo;
```

Comment:
Yeah I noticed that before too -- it's definitely not showing an export named `foo`.